### DISCUSSION TITLE
Can't get PZEM-004T v3 with WEMOS D1 Mini Pro working

### DISCUSSION THREAD

**Discussion by cnoork** [arendst/Tasmota#18170] [Q&A]
I have followed multiple HowTo's but with non of them I get my setup working. I don't get any readings.

![2023-03-12 16_08_40-power02 - Configure Module](https://user-images.githubusercontent.com/17862084/224555232-0460e7ee-77dc-4dd0-95d0-da3688de36bb.jpg)

 I have connected the TX from Wemos to the RX from PZEM-04 and Wemos RX to TX. I have powered the PZEM with 5V and for RX and TX on the PZEM I added the 1k resistor as descripted on multiple location on internet.

![20230312_155517](https://user-images.githubusercontent.com/17862084/224555136-1c2f53f7-b3c3-4fb2-b55b-6acd3f6b3293.jpg)

![20230312_155803](https://user-images.githubusercontent.com/17862084/224555151-7d95cb82-6560-43a5-aed3-ef8d3234ed35.jpg)

And configured Tasmota as followed:

![2023-03-12 16_08_40-power02 - Configure Module](https://user-images.githubusercontent.com/17862084/224555206-1b3143a9-259c-48a3-8c9a-e4e29ef42e0f.jpg)

What could be the reason why my setup is not working. On the the PZEM the led on the power side is on and the led on the RX side is on and also flashing with a higher intensity around every 2 seconds.

**Comment by barbudor** [comment]
There seems to be conflicting templates between the Tasmota Doc site and Tasmota Templates site, and always an ambiguity from driver's authors to where RX/TX are seen from ESP side or the device side...
From the driver's code I see that GPIO shoud be set and connected as
- ESP TX GPIO01 set to "PZEM0xx TX" and connect to PZEM's RX
- ESP RX GPIO03 set to "PZEM0xx RX" and connect to PZEM's TX 

Can you please confirm htis the case ?

Tasmota docs only mention 1 resistor is needed on the RX side and say nothing about TX

**Reply by cnoork** [reply]
Hi barbudor,

Thanks for you reaction. I have indeed connected and configured as you mentioned. Although to be sure I have tried before also the less logic possibilities but off course without any success.

For the second resistor, this one I have also done because with one it didn't work and hopped it solved my issue. In below link you can see that the R8 and R4 is changed, but indeed this is not in all the HowTo's described and probably even not needed. Here the link where I found this: [HOWTO control PZEM-004t #10567](https://github.com/arendst/Tasmota/discussions/10567)

**Reply by cnoork** [reply]
I have just connected a ESP32 but with the same result, no readings

**Reply by pkkrusty** [reply]
I have multiple of these working with the following definitions: ![image](https://user-images.githubusercontent.com/79770016/225156834-ee656247-4ee4-42cc-aae4-ff0cbe06bdc4.png)

**Reply by cnoork** [reply]
pkkrusty, thank you for sharing your config, I expect that the connection from PZEM to Wemos the TX and RX are cross connected. When I was home again I tried this setting but unfortunately it did not solve my issue. I think I will order the USB cable to connect to a PC to check if it will communicate then.

**Reply by cnoork** [reply]
As observation, the LED from TX is blinking, is that the what is to expect?

**Comment by pkkrusty** [comment]
A once-per-second blink of the Rx then immediately the Tx is normal.  A once-per-2-seconds blink of just the Tx (or Rx, can't remember which) is not normal.

Did you build with 
```
#define USE_ENERGY_SENSOR
#define USE_PZEM004T
#define USE_PZEM_AC
```
All my successful builds include all three of those.

https://user-images.githubusercontent.com/79770016/225454342-16a6bfc8-aeec-4239-a4a1-c0f177b560bf.MOV

**Reply by cnoork** [reply]
Hi pkkrusty, thnx for the info. I place the Tasmota through the webpage on the Wemos, no compiling done. I expected this was enabled in the bin file. I will try to compile en do a test again. Or is the sensor.bin including these libraries?

**Reply by pkkrusty** [reply]
Not sure what's included in the sensors.bin. I always compile my own. There's a list in documentation that shows what's included.

**Reply by barbudor** [reply]
If you have the PZEM in the GPIO list, then the feature is included into your binary
PZEM is included in tasmota, tasmota-sensors and tasmota32 standard binaries
https://tasmota.github.io/docs/BUILDS/

**Reply by sfromis** [reply]
Running the tool decode-status will also list most features, if present, including PZEM.

**Comment by pierafal** [comment]
Have anyone solved that? Looks like I have exactly the same issue. I'm using Wemos D1 mini with tasmota and PZEM configuration GPIO1 - 0XX Tx, GPIO3 - 016 Rx (Also tried on different IO's but works the same).
I have 3 PZEM004 (1kOhm soldered between 5V and RX Opto) connected. I'm using 3 diodes BAT42 on PZEM004 TX and 2x4,7kOhm pullup resistors in series. I'm using 3V from Wemos.
When all 3 modules are connected i see 0 values for one device. When i disconnect any 2 devices, Tasmota will show correct values, but only for single PZEM.
I was using ModuleAddress in console with particular PZEM modules connected separatelly. Always I get confirmation {"ModuleAddress":"Done"}. But what's interesting, when I use i.e. ModuleAddress 2, i get the same confirmation when all 3 modules are connected or all modules are disconnected. That's not logical. 
Ezample: I leave only PZEM #2 connected, Tasmota shows power values correctly. Then I'm using ModuleAddress 2, and Tasmota changes values to 0, and after few seconds shows the previous values again.
It looks like the PZEM modules are not addressed correctly, or they do not remember the address set.
What do i do wrong?

**Reply by pierafal** [reply]
Looks like I've found the solution. Seems that Tasmota 12.5.0 that I was using had some issue setting the PZEM address correctly. I've set the addresses using Arduino IDE and Arduino UNO with directly connected PZEM module. Below you'll find a sketch i used. Right now Tasmota displays measures for 3 PZEM modules. Works with BAT42 diodes on TX and 10k pullup resistor as well as without these (but in this case shows some mess at the beginning)
----------------------------------------------------------------------
#include <PZEM004Tv30.h>
PZEM004Tv30 pzem(6,7);
void setup() {
   Serial.begin(9600);
}
// update addr to: 0x01, 0x02, 0x03
uint8_t addr = 0x01;
void loop() {
  Serial.print("Current default address: ");
  Serial.println(pzem.getAddress());
  Serial.print("Current Custom address: ");
  Serial.println(pzem.readAddress());
  pzem.setAddress(addr);
  Serial.print("New address set to: ");
  Serial.println(pzem.readAddress());
  Serial.println();
  delay(10000);
}

**Comment by noega-33** [comment]
comment to cnoork
 [on Mar 12, 2023](https://github.com/arendst/Tasmota/discussions/18170#discussion-4954843)
I had the same issue (not being able to use hardSerial connected to PZEM004) using Lolin wemos D1 mini V4.
I also found that V4 design was  different than original design (using CH340 chip)
in original design, ESP8266 chip RX and TX are connected to Board RX and TX outputs and  connected to TX and RX CH340 pins trough 470 ohm resistors.
in V4 design ESP8266 chip RX pin is directly connected to  CH340 TX pin (no resistor) and ESP8266 chip TX pin is connected to a 470 ohm resistor before TX board output and then connected to CH340 RX pin
I resolved the issue by unsoldering and lifting RX and TX  CH340 pins and connecting 470 ohm resistors as well as removing existing 470 ohm resistor ans replacing it by a small wire.
All work perfectly

**Comment by noega-33** [comment]
V4 design
[Wemos D1 mini V4](https://github.com/user-attachments/assets/e732df31-4afd-47e1-a3ed-dee2ce2879d7)

Original design
![Mod_Wemos_V4](https://github.com/user-attachments/assets/cc63ac71-2146-4658-9d3a-f23bcb7e1965)

My solution
Insert 2 470 resistors between PCB Pads and CH340 chip RX and TX pins
<img width="543" alt="Unsoldering   Lifting RX and Tx pins" src="https://github.com/user-attachments/assets/2491ec66-7473-4265-9c3e-766d41c0fdfa">
<img width="505" alt="Soldering 470 ohm resistors" src="https://github.com/user-attachments/assets/cc340be1-72ce-4ac7-bd0c-4c038561112a">
<img width="420" alt="Connecting Resistors to PCB PAds" src="https://github.com/user-attachments/assets/ee545a4c-b86f-4ad7-b226-f2ca33043d9b">
<img width="425" alt="Final Resistors mod" src="https://github.com/user-attachments/assets/4adf48c9-bb22-4fbc-8d1a-6b15c1e49ce4">
